### DISCUSSION TITLE
Wrong Http Status code while doing stepup Authentication.

### DISCUSSION THREAD

**Discussion by alokkpathak** [OpenIDC/mod_auth_openidc#920] [Q&A] [answered]
Hallo @zandbelt , I implementing the mod_auth_openidc by reading your documents. I am facing issue while doing stepup authentication for our higher login scenario. The issue that we get is the http status code 200 and location instead of 302 and location  in our network calls while doing higherlogin stepup.Here is my OIDC configuration.

 OIDCProviderMetadataURL https://login-ciam-e2e2.whatsappsim.de/signin/oauth2/realms/root/realms/whatsappsim/.well-known/openid-configuration
 OIDCClientID ************
 OIDCClientSecret ***************
 OIDCRedirectURI https://e2e2-apps.whatsappsim.de/apps2mce/services/oidc
 OIDCXForwardedHeaders X-Forwarded-Host X-Forwarded-Port X-Forwarded-Proto
 OIDCCryptoPassphrase "mkXZgN8yiHP4ZgSmHVOS8LD4RDrqnXL90"
 OIDCPKCEMethod S256
 OIDCProviderTokenEndpointAuth client_secret_basic
 OIDCScope "openid portal_care_profile"
 OIDCAuthNHeader X-Forwarded-User
 OIDCPassClaimsAs headers
 OIDCPassIDTokenAs claims payload
 OIDCDefaultLoggedOutURL https://e2e2-apps.whatsappsim.de/apps2mce/services/oidc/brands/whatsappsim/subscriptions_authorized
 OIDCPassCookies mod_auth_openidc_session
 OIDCUnAuthAction auth
 ErrorDocument 401 " "

<Location /apps2mce/services/oidc>
   AuthType openid-connect

    Require claim stackInfo:RAITT
    Require claim stackInfo:Legacy
    Require claim stackInfo:In_Migration
    OIDCUnAuthAction auth
    OIDCPathAuthRequestParams acr_values=password

   <If "%{REQUEST_URI} = '^/apps2mce/services/oidc'">

     RewriteCond %{QUERY_STRING} !stackInfo=RAITT
    RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^RAITT$
    RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=RAITT [L]

    RewriteCond %{QUERY_STRING} !stackInfo=Legacy
    RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^Legacy$
    RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=Legacy [NE,L]

    RewriteCond %{QUERY_STRING} !stackInfo=In_Migration
    RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^In_Migration$
    RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=In_Migration [NE,L]

  </If>

    RewriteCond %{QUERY_STRING} !stackInfo=RAITT
    RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^RAITT$
    RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=RAITT [L]

    RewriteCond %{QUERY_STRING} !stackInfo=Legacy
    RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^Legacy$
    RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=Legacy [NE,L]

   RewriteCond %{QUERY_STRING} !stackInfo=In_Migration
   RewriteCond %{HTTP:OIDC_CLAIM_stackInfo} ^In_Migration$
   RewriteRule ^/?(.*) https://e2e2-apps.whatsappsim.de%{REQUEST_URI}?stackInfo=In_Migration [NE,L]

    LogLevel debug

  </Location>

<Location /apps2mce/services/oidc/higherlogin>
    AuthType openid-connect
    Require claim auth_lvl:20
    OIDCUnAutzAction auth
    OIDCPathAuthRequestParams acr_values=password&prompt=login
    Require valid-user

     LogLevel debug

</Location>

Oidc logs.....
[oidcLogs.txt](https://github.com/zmartzone/mod_auth_openidc/files/9554971/oidcLogs.txt)

I have also attached the image of higher login netwrok calls in chrome browser.
<img width="956" alt="HigherLoginNetworkcalls" src="https://user-images.githubusercontent.com/16973002/189850696-8bdc23c1-0808-4efa-bf4a-7e6d3cd251b2.PNG">

**Comment by zandbelt** [comment, marked as answer]
Step-up for Apache 2.4 does not work with a redirect as Apache does not allow a redirect that late in the process. Instead a HTML page is served that will have a meta refresh tag that redirects the browser for authorization and that page will indeed be served with a status code 200.

**Reply by bthalmayr** [reply]
Thanks for the details @zandbelt , much appreciated.
I'm wondering why the HTTP response heaer 'location' is also set when doing an HTML refresh. Would you mind elaborating on the reasoning?

**Reply by zandbelt** [reply]
that's just the behaviour of meta-equiv refresh see https://en.wikipedia.org/wiki/Meta_refresh

**Reply by bthalmayr** [reply]
Hmm, HTML based refresh can not set an HTTP response header, can it?

I thought there are only 3 ways for redirection, HTTP redirect (send HTTP status code starting with '3' and HTTP response header 'location'), HTML refresh , JavaScript code. At least this seems to be the case when reading [https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections)

**Reply by zandbelt** [reply]
ok, after closer inspection it appears that the code actually sets the location header before generating the HTML refresh page

**Reply by bthalmayr** [reply]
Thanks for checking @zandbelt . So either this happens by accident or by intention as you want to give non-browser based client the chance to leverage an HTTP response header instead of parsing HTML content. At least I was a bit astonishing to see this behavior.

**Comment by alokkpathak** [comment]
Thanks @zandbelt for your response.
do we have any workaround for this to provide the status code 302 to our frontend?

**Reply by zandbelt** [reply]
no, the refresh will eventually lead to a redirect as mentioned above
